### ISSUE TITLE
Versioning with tags diverging from master

### ISSUE THREAD

**Issue by AbdealiLoKo** [pypa/setuptools-scm#330]
I have a scenario where I have my backported a bugfix and have a branch which I have tagged (this is an example graph):

```
commit1 (tagged as v0.3.0) -> commit2 ------------------> commit4 -----> commit5
                                  |
                                  +-----> commit3 (tagged as v0.4.0)
```

when I am on commit5 and do `python setup.py --version` it shows me `0.3.1.dev501` while I was expecting it to show `v0.4.1.devX` as I have already released `v0.4.0`

Am I setting something wrong that it is not detecting the v0.4.0 tag? 
I just use:
```
    use_scm_version={
        'write_to': 'mypkg/version.txt',
    },
```

**Comment by RonnyPfannschmidt** [comment]
its as expected, commit 5 is not a ancestor of the tagged commit3 - the next tag that can be seen from commit 4 is the `v0.3.0` one

**Comment by AbdealiLoKo** [comment]
How would such cases be handled though ?
Isnt it a common practice to cut out a v0.3.x branch and start testing it. 
And when any bugs are found its fixed in 0.3.x and released? 

Should I have marked commit2 as v0.4.0rc1 ?
I'm trying to understand how/if setuptools_scm handles this and if I'm not following a best practice - I hope to start following it

**Comment by RonnyPfannschmidt** [comment]
im not sure i follow,
setuptools_scm naturally handles the cases you described as far as i understood,

the problem i see in your case is that your `v0.4.0` is on a branch that hasn't been merged

so the "mainline" as far as i understood is a branch following `v0.3.x`

the correct way as far as  i understood it, would be to merge

however your perspective on the branches/versions and your intent are rather unclear to me,
i suspect a slight misunderstanding/X-Y problem

**Comment by AbdealiLoKo** [comment]
Yeah, let me describe our release process a bit better.
Our process works as follows:

 - Developers keep adding fixes and features into the master branch via PRs.
 - On day X we decide we have sufficient features to make a v0.4.0 release. So, a release manager is allocated and he/she makes a branch called v0.4.x
 - The v0.4.x branch is given to anyone willing to test the release and they test the functionality to ensure things haven't broken (meanwhile master continues as normal with feature and bugfix merges)
 - If any blocker bugs are found in v0.4.x - it is reported as an issue. And two scenarios play out: Either it is already fixed in master (after v0.4.x was branched out) or it is not. If not fixed - one of the developers (or release manager) creates a PR on master to fix the bug.
 - The 0.4.0 release manager backports this commit - normally by a cherry-pick and appropriately fixing conflicts - to fix it in v0.4.x branch
 - Finally when v0.4.x branch is OKed by testers (or a week passes), the commit in v0.4.x is tagged as v0.4.0

Notes:
 - All bugfixes will always be pushed to master and then backported into v0.4.x branch as necessary
 - The v0.4.x branch is never merged back to master - as master will have diverged and that bugfix was taken form master anyway. Hence the commit tagged as v0.4.1 will never be in master
 - When we next need to release a v0.4.1 - the same v0.4.x branch is taken and more bugfixes are backported to it and a v0.4.1 tag is created (again, the commit tagged as v0.4.1 won't be merged with master)

**Comment by RonnyPfannschmidt** [comment]
am i correctly understanding that at certain points in time you branch of release-series from master, and then master should prepare the next release-series while the release-series branch should handle more bug-fixes on those releases


for those cases i believe putting a specific dev tag onto the mainlines could sort it out in a relatively approachable manner

aka something like 
```
A
| \
B  X
|  |
C  Y

TAGS:

B: v0.5.0.dev
X: v0.4.0
Y: v0.4.1
```

i haven't tried this for myself due to lack of time, please experiment in a separate repo if such a setup works out for you - if it doesn't there are still various options to take branch and close tags into account

**Comment by AbdealiLoKo** [comment]
Ah. I see. If I create a tag at `v0.6.0.dev` or `v0.6.0.dev0` it works fine.
I had tried by setting `v0.6.0.dev1` as I wasnt aware that setting it to 0 was a valid version.
I think this resolves my issue - whenever we make a fork for the release-series branch - I would just do a `.dev0` tag.

And putting it at the point that is after the v0.4.x branch is taken out - will ensure that the next version will always be `v0.5.1.dev2g...` and hence will be considered higher than anything in the v0.4.x branch - which is the correct behaviour.

Closing the issue as I think this solves my question. Thanks!
